File: imu_calib_matlab/src/analysis/analyze_allan_common.m
function allan = analyze_allan_common(t, x, sensorType, varargin)
%ANALYZE_ALLAN_COMMON Shared Allan deviation implementation for 3-axis signals.

validateattributes(t, {'numeric'}, {'column', 'nonempty', 'finite'}, mfilename, 't', 1);
validateattributes(x, {'numeric'}, {'2d', 'ncols', 3, 'nonempty', 'finite'}, mfilename, 'x', 2);
if size(x, 1) ~= numel(t)
    error('analyze_allan_common:LengthMismatch', ...
        't and x must have matching lengths.');
end

opts = parse_options(varargin{:});
N = size(x, 1);
dt = median(diff(t));

allan = struct();
allan.valid = false;
allan.sensor_type = sensorType;
allan.message = '';
allan.tau = [];
allan.adev = [];
allan.num_samples = N;

if N < opts.min_samples
    allan.message = opts.validity_message_if_short;
    return;
end

maxM = max(2, floor(N / 10));
if maxM < 2
    allan.message = 'Insufficient data length for Allan deviation.';
    return;
end

switch lower(opts.tau_mode)
    case 'logspace'
        mList = unique(max(1, round(logspace(0, log10(maxM), opts.num_tau))));
    otherwise
        mList = 1:maxM;
end

mList = mList(mList >= 1);
tau = mList(:) * dt;
adev = NaN(numel(mList), 3);

for k = 1:numel(mList)
    m = mList(k);
    clusterCount = floor(N / m);
    if clusterCount < 2
        continue;
    end

    trimmed = x(1:(clusterCount * m), :);
    clustered = squeeze(mean(reshape(trimmed, m, clusterCount, 3), 1));
    if clusterCount == 2 && isvector(clustered)
        clustered = reshape(clustered, 2, 3);
    end

    diffCluster = diff(clustered, 1, 1);
    adev(k, :) = sqrt(0.5 * mean(diffCluster .^ 2, 1));
end

validRows = all(isfinite(adev), 2);
tau = tau(validRows);
adev = adev(validRows, :);

if isempty(tau)
    allan.message = 'Allan deviation could not be estimated from the provided data.';
    return;
end

allan.valid = true;
allan.message = 'Allan deviation estimated using non-overlapping cluster averages.';
allan.tau = tau;
allan.adev = adev;
allan.estimate = estimate_allan_params(tau, adev, sensorType);
end

function opts = parse_options(varargin)
defaults = default_calib_options();
opts = defaults.allan;

if mod(numel(varargin), 2) ~= 0
    error('analyze_allan_common:InvalidInput', ...
        'Optional arguments must be provided as name/value pairs.');
end

for k = 1:2:numel(varargin)
    name = lower(char(varargin{k}));
    value = varargin{k + 1};
    switch name
        case 'min_samples'
            opts.min_samples = double(value);
        case 'num_tau'
            opts.num_tau = double(value);
        case 'tau_mode'
            opts.tau_mode = char(value);
        case 'validity_message_if_short'
            opts.validity_message_if_short = char(value);
        otherwise
            error('analyze_allan_common:UnknownOption', ...
                'Unknown option "%s".', char(varargin{k}));
    end
end
end

function estimate = estimate_allan_params(tau, adev, sensorType)
estimate = struct();
estimate.noise_density = adev(1, :).';
estimate.bias_instability = min(adev, [], 1).';
estimate.random_walk = (adev(end, :) ./ sqrt(tau(end))).';
estimate.confidence = 'low_to_medium';
estimate.notes = sprintf(['Stage-1 %s Allan estimates are coarse summaries intended for ' ...
    'engineering inspection, not high-accuracy metrology.'], sensorType);
end

File: imu_calib_python/imu_calib/calib/analyze_allan_common.py
from __future__ import annotations

import numpy as np

from imu_calib.runtime.default_calib_options import default_calib_options
from imu_calib.utils.exceptions import ImuCalibError
from imu_calib.utils.math_utils import as_matrix_n3, validate_time_vector


def analyze_allan_common(
    t: np.ndarray,
    x: np.ndarray,
    sensor_type: str,
    *,
    min_samples: int | None = None,
    num_tau: int | None = None,
    tau_mode: str | None = None,
    validity_message_if_short: str | None = None,
) -> dict:
    """Shared Allan deviation implementation for 3-axis signals."""
    t = validate_time_vector(t, "t")
    x = as_matrix_n3(x, "x")
    if x.shape[0] != t.size:
        raise ImuCalibError("t and x must have matching lengths.")

    defaults = default_calib_options()["allan"]
    min_samples = defaults["min_samples"] if min_samples is None else int(min_samples)
    num_tau = defaults["num_tau"] if num_tau is None else int(num_tau)
    tau_mode = defaults["tau_mode"] if tau_mode is None else str(tau_mode)
    validity_message_if_short = (
        defaults["validity_message_if_short"]
        if validity_message_if_short is None
        else str(validity_message_if_short)
    )

    N = x.shape[0]
    dt = float(np.median(np.diff(t)))
    allan = {
        "valid": False,
        "sensor_type": sensor_type,
        "message": "",
        "tau": np.array([], dtype=float),
        "adev": np.empty((0, 3), dtype=float),
        "num_samples": int(N),
    }

    if N < min_samples:
        allan["message"] = validity_message_if_short
        return allan

    max_m = max(2, N // 10)
    if max_m < 2:
        allan["message"] = "Insufficient data length for Allan deviation."
        return allan

    if tau_mode.lower() == "logspace":
        m_list = np.unique(np.maximum(1, np.round(np.logspace(0, np.log10(max_m), num_tau)).astype(int)))
    else:
        m_list = np.arange(1, max_m + 1, dtype=int)

    tau = m_list.astype(float) * dt
    adev = np.full((m_list.size, 3), np.nan, dtype=float)
    for k, m in enumerate(m_list):
        cluster_count = N // m
        if cluster_count < 2:
            continue
        trimmed = x[: cluster_count * m, :]
        clustered = trimmed.reshape(cluster_count, m, 3).mean(axis=1)
        diff_cluster = np.diff(clustered, axis=0)
        adev[k, :] = np.sqrt(0.5 * np.mean(diff_cluster**2, axis=0))

    valid_rows = np.all(np.isfinite(adev), axis=1)
    tau = tau[valid_rows]
    adev = adev[valid_rows, :]
    if tau.size == 0:
        allan["message"] = "Allan deviation could not be estimated from the provided data."
        return allan

    allan["valid"] = True
    allan["message"] = "Allan deviation estimated using non-overlapping cluster averages."
    allan["tau"] = tau
    allan["adev"] = adev
    allan["estimate"] = _estimate_allan_params(tau, adev, sensor_type)
    return allan


def _estimate_allan_params(tau: np.ndarray, adev: np.ndarray, sensor_type: str) -> dict:
    return {
        "noise_density": adev[0, :].copy(),
        "bias_instability": np.min(adev, axis=0),
        "random_walk": adev[-1, :] / np.sqrt(tau[-1]),
        "confidence": "low_to_medium",
        "notes": (
            f"Stage-2 {sensor_type} Allan estimates are coarse summaries intended "
            "for engineering inspection, not high-accuracy metrology."
        ),
    }
